rat/compiler.py

```diff
@@ -523,13 +580,58 @@ def _generate_python_source(self, ordered_expr_trees: List[ops.Expr]):
                 code += f"        {constrained_reference} = jax.numpy.pad({constrained_reference}, (0, {variable.padded_size() - size}))\n"
         code += "\n    return total, parameters\n"
 
+        ordered_expr_trees = list(reversed(self.expr_tree_list))
+
         # Transform parameters
-        code += "\ndef transform_parameters(data, subscripts, parameters):"
+        code += "\ndef transform_parameters(data, subscripts, first_in_group_indicators, parameters):"
         for top_expr in ordered_expr_trees:
-            if isinstance(top_expr, ops.Assignment):
-                lhs_key = top_expr.lhs.get_key()
-                code += f"\n    # {lhs_key}\n"
-                code += f"    parameters['{lhs_key}'] = {top_expr.rhs.code()}\n"
+            match top_expr:
+                case ops.Assignment(lhs, rhs):
+                    # Assume left hand side is a parameter
+                    lhs_key = top_expr.lhs.get_key()
+                    code += f"\n    # {lhs_key}\n"
+
+                    lhs_size = lhs.variable.size()
+                    lhs_is_scalar = lhs_size == None
+                    lhs_used_in_rhs = False
+                    need_first_in_group_indicator = False
+                    if not lhs_is_scalar:
+                        carry_size = 0
+                        # Because left hand side is parameter, we only need to search for parameters on the right hand side
+                        for symbol in ops.search_tree(rhs, ops.Param):
+                            symbol_key = symbol.get_key()
+
+                            # Save an indicator if the left hand side is used on the right hand side
+                            if symbol_key == lhs_key:
+                                lhs_used_in_rhs = True
+
+                                # Figure out the carry size needed
+                                shifts = symbol.subscript.shifts
+
+                                carry_size = max(carry_size, shifts[-1])
+
+                                # There will be groupings if there is more than one shift
+                                if len(shifts) > 1:
+                                    need_first_in_group_indicator = True
+
+                    if lhs_used_in_rhs:
+                        # We need to scan in this case
+                        scan_function_name = f"scan_function_{self._get_unique_identifier()}"
```

Comment:
lol that's alot of cryptic looking fstrings

Reply:
I wonder if we can make the code generation code tidier. With the indents and stuff it looks hard to read and keep track.

Reply:
Lol yeah at the f-strings. I should probably add some example comments. Much harder to read than what they actually are.

I agree this code feels kinda at its limits. I'm not quite sure how to do the code-gen more cleanly though, so my thought is to just wait until this code fails to do anything more complex (since I'm not sure in what direction to take it). If you can think of a way to improve, go for it.